NoClassDefFoundError for NonBlockingIdentityHashMap when running an application with BTrace

When running an application with BTrace, the following error occurs:

java.lang.NoClassDefFoundError: org/openjdk/btrace/libs/org/jctools/maps/NonBlockingIdentityHashMap
	at org.openjdk.btrace.instr.ClassCache.<init>(ClassCache.java:80)
	at org.openjdk.btrace.instr.ClassCache$Singleton.<clinit>(ClassCache.java:158)
	at org.openjdk.btrace.instr.ClassCache.getInstance(ClassCache.java:90)
	at org.openjdk.btrace.instr.InstrumentUtils.collectHierarchyClosure(InstrumentUtils.java:78)
	at org.openjdk.btrace.instr.InstrumentUtils.collectHierarchyClosure(InstrumentUtils.java:60)
	at org.openjdk.btrace.instr.BTraceClassWriter.getCommonSuperClass(BTraceClassWriter.java:111)
	at org.openjdk.btrace.libs.org.objectweb.asm.SymbolTable.addMergedType(SymbolTable.java:1202)
	at org.openjdk.btrace.libs.org.objectweb.asm.Frame.merge(Frame.java:1299)
	at org.openjdk.btrace.libs.org.objectweb.asm.Frame.merge(Frame.java:1197)
	at org.openjdk.btrace.libs.org.objectweb.asm.MethodWriter.computeAllFrames(MethodWriter.java:1611)
	at org.openjdk.btrace.libs.org.objectweb.asm.MethodWriter.visitMaxs(MethodWriter.java:1547)
	at org.openjdk.btrace.libs.org.objectweb.asm.tree.MethodNode.accept(MethodNode.java:767)
	at org.openjdk.btrace.libs.org.objectweb.asm.tree.MethodNode.accept(MethodNode.java:647)
	at org.openjdk.btrace.libs.org.objectweb.asm.tree.ClassNode.accept(ClassNode.java:468)
	at org.openjdk.btrace.instr.BTraceProbeNode.getBytecode(BTraceProbeNode.java:252)
	at org.openjdk.btrace.instr.BTraceProbeNode.getFullBytecode(BTraceProbeNode.java:210)
	at org.openjdk.btrace.instr.BTraceProbePersisted.<init>(BTraceProbePersisted.java:83)
	at org.openjdk.btrace.instr.BTraceProbePersisted.from(BTraceProbePersisted.java:90)
	at org.openjdk.btrace.instr.InstrPackGenerator.generateProbePack(InstrPackGenerator.java:19)
	at org.openjdk.btrace.compiler.CompilerHelper.compile(CompilerHelper.java:122)
	at org.openjdk.btrace.compiler.Compiler.compile(Compiler.java:244)
	at org.openjdk.btrace.compiler.Compiler.compile(Compiler.java:234)
	at org.openjdk.btrace.compiler.Compiler.main(Compiler.java:172)
Caused by: java.lang.ClassNotFoundException: org.openjdk.btrace.libs.org.jctools.maps.NonBlockingIdentityHashMap
	at java.net.URLClassLoader.findClass(URLClassLoader.java:382)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:418)
	at sun.misc.Launcher$AppClassLoader.loadClass(Launcher.java:355)
	at java.lang.ClassLoader.loadClass(ClassLoader.java:351)
	... 23 more

Reason seems to be that ClassCache does the following imports
- import org.openjdk.btrace.libs.org.jctools.maps.NonBlockingHashMap;
- import org.openjdk.btrace.libs.org.jctools.maps.NonBlockingIdentityHashMap;
which are however not part of btrace-client-2.2.3.jar

--- a/btrace-dist/build.gradle
+++ b/btrace-dist/build.gradle
@@ -144,8 +144,16 @@ task clientJar(type: ShadowJar) {
     exclude 'org/checkerframework/**'
     exclude 'org/codehaus/**'
 
-    exclude 'org/jctools/maps/**'
-    exclude 'org/jctools/util/**'
+    exclude {
+        it.path.startsWith('org/jctools/maps/') &&
+        !it.path.startsWith("org/jctools/maps/NonBlockingIdentityHashMap") &&
+        !it.path.startsWith("org/jctools/maps/NonBlockingHashMap")
+    }
+    exclude {
+        // 'org/jctools/util/**'
+        it.path.startsWith('org/jctools/util/') &&
+        !it.path.startsWith("org/jctools/util/UnsafeAccess")
+    }
     exclude 'META-INF/services/com.sun.*'
     exclude 'META-INF/services/javax.annotation.*'
 

